harness/determined/launch/wrap_rank.py

```diff
@@ -0,0 +1,92 @@
+"""
+wrap_rank.py prefixes every line of output from a worker process with the rank
+that emitted it.
+
+In distributed training, the rank prefix added by wrap_rank.py is necessary for
+the WebUI log viewer's filter-by-rank feature to work.
+
+Additionally, when used in a Determined container, wrap_rank.py redirects stdout
+and stderr of the worker process to the stdout and stderr of the container.  The
+purpose of this is to save network bandwidth when launchers like mpirun or
+horovodrun are used, as they often are configured to send all logs from worker
+nodes to the chief node over the network.  This may be disabled with the
+``--no-redirect-stdio`` flag.
+"""
+import argparse
+import contextlib
+import os
+import subprocess
+import sys
+import threading
+from typing import BinaryIO, List
+
+from determined import constants
+
+
+def forward_stream(src_stream: BinaryIO, dst_stream: BinaryIO, rank: str) -> None:
+    for line in iter(src_stream.readline, b""):
+        line = f"[rank={rank}] ".encode() + line
+        os.write(dst_stream.fileno(), line)
+
+
+def run_all(ts: List[threading.Thread]) -> None:
+    for t in ts:
+        t.start()
+    for t in ts:
+        t.join()
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        usage="wrap_rank.py [-h] [--no-redirect-stdio] RANK SCRIPT...",
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument("--no-redirect-stdio", action="store_true")
+    parser.add_argument(
+        "rank",
+        metavar="RANK",
+        help=(
+            "Can be an integer rank or the name of an "
+            "environment variable containing an integer rank."
+        ),
+    )
+    parser.add_argument(
+        "script", nargs=argparse.REMAINDER, metavar="SCRIPT...", help="The worker command."
+    )
+    args = parser.parse_args()
+
+    if set("0123456789") >= set(args.rank):
```

Comment:
lol, this is neat

Reply:
just recently learned this trick from @mpkouznetsov